### PR TITLE
feat(perf): Cache commutative instructions in constant folding

### PR DESCRIPTION
**Pull request by vezenovm**
# Description

## Problem\*

Resolves 

## Summary\*

Keeping a draft as I want to see how it runs in the code size benchmarks 

## Documentation\*

Check one:
- [ ] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by jfecher** [comment]
I don't understand the regression_5252 size increase

**Comment by vezenovm** [comment]
> I don't understand the regression_5252 size increase

Yeah I don't either. I guess we can close this PR as it doesn't seem to be providing many benefits.

File: compiler/noirc_evaluator/src/ssa/opt/constant_folding.rs [hunk 1]
```diff
@@ -264,18 +264,62 @@ impl Context {
...
+                | BinaryOp::Mul
+                | BinaryOp::Eq
+                | BinaryOp::And
+                | BinaryOp::Or
+                | BinaryOp::Xor => {
+                    let commutative_instruction = Instruction::Binary(Binary {
+                        lhs: binary.rhs,
+                        rhs: binary.lhs,
+                        operator: binary.operator,
+                    });
```
[the remark this answers is not in the document]
Comment (vezenovm) [reply compiler/noirc_evaluator/src/ssa/opt/constant_folding.rs]: Done. However, we seem to be getting similar results as before.

File: compiler/noirc_evaluator/src/ssa/opt/constant_folding.rs [hunk 2]
```diff
@@ -192,8 +192,32 @@ impl Context {
         }
 
         // Resolve any inputs to ensure that we're comparing like-for-like instructions.
+        let instruction = instruction
+            .map_values(|value_id| resolve_cache(dfg, constraint_simplification_mapping, value_id));
+
+        // Sort operands of commutative instructions
+        if let Instruction::Binary(binary) = &instruction {
+            match binary.operator {
```
Comment (jfecher) [inline compiler/noirc_evaluator/src/ssa/opt/constant_folding.rs:200]: Oh, I meant sorting when the instructions are first created. Not during constant folding.

**Comment by vezenovm** [comment]
Going to close this PR as the benefits seem minimal in Brillig code size and actually cause regressions for ACIR.

**Pull request closed without merging**